### ISSUE TITLE
Ideia para mudar estagio dos leads

### ISSUE THREAD

**Issue by sergiors** [sergiors/rdstation#1]
Estou testando achar nomes que transmitem melhor as ações, até o momento ficaram isso.

`Sergiors\RDStation\Lead\ChangesToClient`
`Sergiors\RDStation\Lead\ChangesToQualified`

@rfdeoliveira ideias?

**Comment by ronaldooliveiradevbr** [comment]
@sergiors pra mim tá bem claro. Como alternativas, pensei nestas possibilides:

> Sergiors\RDStation\Lead\StageToClient

> Sergiors\RDStation\Lead\StageToQualified

E a respeito de oportunidade que tal:

> Sergiors\RDStation\Lead\TrueOpportunity

> Sergiors\RDStation\Lead\FalseOpportunity ?

**Comment by sergiors** [comment]
Opa, bem melhor `StageToClient` e `StageToQualified`.

Sobre `Opportunity`, acho que deveria ser flag, pelo que vi na API, toda mudança de lead vc pode definir como `opportunity = true` ou `false`

Outro pensamento e ter o objeto `Lead` e passamentos outro objeto pra ele, com o comportamento de `Stage`,

**Comment by ronaldooliveiradevbr** [comment]
Verdade! `Opportunity` nem precisa ser objeto, acho que como atributo é OK.

Bacana a ideia do `Stage`. Faria sentido o contrário? Exemplo:

`$stage = new StageToClient($lead);`
`$stage->trigger();`

o detalhe é que aí seria necessária a consulta do `Lead`, imagino

**Comment by sergiors** [comment]
Acho que ficaria trabalhoso, pois dentro da trigger do `Lead` instanciamos o `Request`...
o `StageToClient` que escreveria esses comportamentos e o `Lead` cuida dele..

Mas seria bacana implementar antes o código para ter uma noção antes de descartar a ideia/selecionar.

**Comment by ronaldooliveiradevbr** [comment]
Vou fazer um teste e tentar abrir um PR hoje. 

Cara, pensei no que você falou de passar o `Stage` para o `Lead`, me explica melhor essa ideia pra eu começar a implementar, por favor =)

Como seria essa passagem? Seria algo assim?

```php

interface StageLifecycle {}

class Client implements StageLifecycle {}

$lead->stageTo(new Client());

...

// no Lead
public function stageTo(StageLifecycle $stage)
{
    ...

    $stage->trigger();
}

```

Valeu, Abraço!

**Comment by sergiors** [comment]
Oi @rfdeoliveira, irei analisar com calma hoje a noite, mas o `Stage` teria q ser um `Lead`, afinal é apenas um outro estágio do `Lead`, então ele continua sendo Lead, mas em outro estágio..
Talvez uma abordagem de composição de tipos.

**Comment by ronaldooliveiradevbr** [comment]
Boa! Quando tiver uma ideia, traz aí pra discutirmos o/
